Features/bschwarzent/24.2/286465 disable navigatedownbutton

bschwarzent opened the pull request.

bschwarzent committed:
DesktopBench: cleanup code

- Use scout.assertParameter() instead of custom Error
- Use 'page' as parameter name instead of 'node'
- Adjust return type from 'Form | Table' to 'OutlineContent'
- Extract method

bschwarzent committed:
PageWithTable: set 'childrenLoaded' property in loadChildren()

- Set the property 'childrenLoaded' (which is used used by
  TreeNode#ensureLoadChildren) in the method loadChildren(). The same
  pattern is used in PageWithNode.
- Adjust jsdoc of Tree#loadChildren.
- Use then() instead of done() for consistency reasons.

bschwarzent committed:
Outline: harmonize drill-down logic

- Move similar logic to a new method: Outline#drillDown
- Add spec

bschwarzent committed:
NavigateButton: add jsdoc

bschwarzent committed:
NavigateDownButton: automatically disable when table is empty

Two new methods were added to NavigateButton: _hasDetailForm() and
_hasDetailTable(). They return the "availability" of the corresponding
detail content element, but _hasDetailTable() by default now also
returns false if the detail table is available but has no rows.

This allows the NavigateDownButton to be automatically disabled when
the detail table is empty. This is consistent with the disabled state
when no row is selected. As a consequence it is no longer necessary to
explicitly set detailTableVisible to false to disable the
NavigateDownButton.

The NavigateDownButton relies on the content of the detail table. For
table pages the detail table is naturally up-to-date, because it is the
source of the data. Child nodes are always created for a specific row.
Node pages on the other hand are defined by their child nodes. The
detail table is a secondary view that only mirrors the child pages, it
cannot create new rows. By default, the detail table is rebuilt during
the loadChildren() method. But that does not happen if child nodes are
added or removed dynamically. To automatically reflect the changes in
the child nodes in the detail table, we add a local tree listener to
the outline. Whenever the tree is changed, we determine the parent node
of the affected nodes and inform them via outline mediator that they
should rebuild their detail table.

286465

bschwarzent committed:
Add TS interfaces to improve IDE support

- Table/Tree implements Filterable
- TableRow/TreeNode implements FilterElement
- LazyNodeFilter implements Filter

bschwarzent committed:
Scrolling: improve visibility of expanded tree nodes

- When a tree node is selected, ensureExpansionVisible() is called to
  make sure the expanded children of this node are within the viewport.
- This method should only consider nodes that are actually visible, i.e.
  not filtered (e.g. by lazy expansion).
- If a node has no children, at least the node itself should be scrolled
  inside the viewport.
- Because the outline tree has no horizontal scrollbar, no scrolling
  should be performed in that direction (even when the tree nodes are
  pushed out of the visible area).

336993

bschwarzent committed:
Add utility functions

- dates.formatTimestamp(): format a date in a fixed timestamp format
- numbers.randomInt(): generate a random integer

bschwarzent committed:
PageWithTable: allow rows without child page

In Java, it is possible to have rows in table pages without a
corresponding child node. The same should be possible in JS.

- Allow null as result in _createChildPageInternal()
- Automatically filter rows without pages in pagesForTableRows()
  and updatePagesFromTableRows().
- The unnecessary method pageForTableRow() was deprecated. It offers
  no advantage over accessing the TableRow#page field directly.

bschwarzent committed:
router: remove unnecessary warning "no route registered for null"

The initial location has no explicit route, but this is not an error.
Because the URL is not only used for routing (e.g. we can have query
params), we cannot reliably determine when to emit a warning. To avoid
unnecessary log entries, the log level of the warning was reduced to
info.

bschwarzent committed:
objects: remove old, unused development utility

objects.checkFunctionOverrides() was used in the pre-ES6 era to check
in the F12 developer console if the signature of "overwritten" methods
in the prototype chain matches. This utility no longer works correctly
and is no longer needed (we now rely on ES6/TS classes and IDE support).

bschwarzent committed:
PageWithNodes: add 'reloadable' flag

PageWithNodes are not reloadable by default, because the child nodes
(if any) are typically created statically. However, when child pages
are created programmatically by implementing _createChildPages(), the
new 'reloadable' flag can be set to true.

Before adding new child pages, the obsolete child pages have to be
removed. Nodes that already exist in the tree should remain untouched,
because removing and re-inserting them would destroy them. They will
be ignored by insertNodes().

cguglielmo commented on eclipse-scout-core/src/util/dates.ts:
```diff
@@ -465,6 +465,43 @@ export const dates = {
+   *
+   * @param options A {@link DatesFormatTimestampOptions} object or a date.
+   */
+  formatTimestamp(options?: DatesFormatTimestampOptions | Date): string {
```
What is the difference to scout.dates.format(new Date(), scout.Locale.DEFAULT, 'yyyy-MM-dd hh:mm:ss.SSS')?
To omit the time or millisecons, you can just use an adjusted pattern, so it does not need a custom implementation. 
The name is also a bit confusing, because it does not reflect the difference to format(). Maybe formatNormalized, or basicFormat, simpleFormat or something like it.

bschwarzent replied:
Aggred, it is too confusing. I reverted this changed. Instead, the `dates.format()` function can be used as suggested.

Things to note:
- `Locale.DEFAULT` cannot be used as second argument, because that constant is only a _LocaleModel_, not a _Locale_. I used `Locale.ensure()` instead.
- The pattern should use uppercase `HH` for the hours to make sure DateFormat uses the 24-hour format (0-23) rather than the 12-hour format (1-12).

cguglielmo commented on eclipse-scout-core/src/desktop/outline/navigation/NavigateDownButton.ts:
```diff
@@ -61,30 +67,32 @@ export class NavigateDownButton extends NavigateButton {
+      let selectedRows = this.node.detailTable.selectedRows;
+      // Drill down is only possible when a single row is selected and that row is linked to a page
+      if (selectedRows.length === 1 && selectedRows[0].page) {
+        return this.node.pageForTableRow(selectedRows[0]);
```
Why don't you just use selectedRows\[0].page? Why do we even need this method? :)

bschwarzent replied:
You are correct, this method is not necessary. I marked it as deprecated and replaced all usages. The property "page" on `TableRow` now has a short js doc.

cguglielmo commented on eclipse-scout-core/src/desktop/outline/Outline.ts:
```diff
@@ -159,14 +161,36 @@ export class Outline extends Tree implements DisplayParent, OutlineModel {
     return scout.create(OutlineMediator);
   }
 
+  protected _installLocalTreeListener() {
```
How can the node not be a Page? I would skip the filter and just add a cast.

bschwarzent replied:
Done.

cguglielmo commented on eclipse-scout-core/src/desktop/outline/Outline.ts:
```diff
@@ -159,14 +161,36 @@ export class Outline extends Tree implements DisplayParent, OutlineModel {
+    });
+
+    // Mediate subtree changes
+    let handleSubTreeChanged = (nodes: TreeNode | TreeNode[]) => {
```
Can you explain in the commit message, what the purpose of this change is? Is this relevant for the navigate down button?

bschwarzent replied:
A more detailed description was added.

bschwarzent committed:
Page: remove destroyed detail form and table from page

A page may reference a detailForm or a detailTable. If the corresponding
widget is destroyed, the member should also be set to null. Otherwise,
it may happen that the destroyed widget is still accessed, creating the
potential for unexpected errors ('Widget is destroyed').

For example, if a node page with both a detail form and a detail table
is destroyed while still being selected, the DesktopBench will be
triggered by a pageChanged event and tries to update the outline content
with an already destroyed widget.
